// happy-dom/lib/nodes/html-option-element/HTMLOptionElement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const HTMLElement_1 = __importDefault(require("../html-element/HTMLElement"));
/**
 * HTML Option Element.
 *
 * Reference:
 * https://developer.mozilla.org/en-US/docs/Web/API/HTMLOptionElement.
 */
class HTMLOptionElement extends HTMLElement_1.default {
    /**
     * Returns inner text, which is the rendered appearance of text.
     *
     * @returns Inner text.
     */
    get text() {
        return this.innerText;
    }
    /**
     * Sets the inner text, which is the rendered appearance of text.
     *
     * @param innerText Inner text.
     */
    set text(text) {
        this.innerText = text;
    }
    /**
     * Returns index.
     *
     * @returns Index.
     */
    get index() {
        return this._index;
    }
    /**
     * Returns the parent form element.
     *
     * @returns Form.
     */
    get form() {
        let parent = this.parentNode;
        while (parent && parent.tagName !== 'FORM') {
            parent = parent.parentNode;
        }
        return parent;
    }
    /**
     * Returns selected.
     *
     * @returns Selected.
     */
    get selected() {
        const parentNode = this.parentNode;
        if (parentNode?.tagName === 'SELECT') {
            let index = -1;
            for (let i = 0; i < parentNode.options.length; i++) {
                if (parentNode.options[i] === this) {
                    index = i;
                    break;
                }
            }
            return index !== -1 && parentNode.options.selectedIndex === index;
        }
        return false;
    }
    /**
     * Sets selected.
     *
     * @param selected Selected.
     */
    set selected(selected) {
        const parentNode = this.parentNode;
        if (parentNode?.tagName === 'SELECT') {
            if (selected) {
                let index = -1;
                for (let i = 0; i < parentNode.options.length; i++) {
                    if (parentNode.options[i] === this) {
                        index = i;
                        break;
                    }
                }
                if (index !== -1) {
                    parentNode.options.selectedIndex = index;
                }
            }
            else if (parentNode.options.length) {
                parentNode.options.selectedIndex = 0;
            }
            else {
                parentNode.options.selectedIndex = -1;
            }
        }
    }
    /**
     * Returns disabled.
     *
     * @returns Disabled.
     */
    get disabled() {
        return this.getAttributeNS(null, 'disabled') !== null;
    }
    /**
     * Sets disabled.
     *
     * @param disabled Disabled.
     */
    set disabled(disabled) {
        if (!disabled) {
            this.removeAttributeNS(null, 'disabled');
        }
        else {
            this.setAttributeNS(null, 'disabled', '');
        }
    }
    /**
     * Returns value.
     *
     * @returns Value.
     */
    get value() {
        return this.getAttributeNS(null, 'value') || '';
    }
    /**
     * Sets value.
     *
     * @param value Value.
     */
    set value(value) {
        this.setAttributeNS(null, 'value', value);
    }
}
exports.default = HTMLOptionElement;
